File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/OrganisePlayerCardsCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.{BurracoGameInitiated, BurracoGameInitiatedTurnEnd, BurracoGameInitiatedTurnExecution, BurracoGameInitiatedTurnStart}
import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.decks.Card
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class OrganisePlayerCardsCmd(
                                gameIdentity: GameIdentity,
                                playerIdentity: PlayerIdentity,
                                orderedCards: List[Card]
                              ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class OrganisePlayerCardsHandler(
                                  gameRepositoryPort: GameRepositoryPort
                                )(implicit val ec: scala.concurrent.ExecutionContext)
  extends CommandHandler[OrganisePlayerCardsCmd] {

  override def handleAsync(command: OrganisePlayerCardsCmd): Future[Unit] = {
    for {
      gameInitialised <- gameRepositoryPort.findBurracoGameInitialisedBy(gameIdentity = command.gameIdentity)
      gameInitialisedUpdated = reorderPlayerCards(gameInitialised, command.playerIdentity, command.orderedCards)
    } yield gameRepositoryPort.save(gameInitialisedUpdated)
  }

  override def handle(command: OrganisePlayerCardsCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val gameInitialised = Await.result(gameRepositoryPort.findBurracoGameInitialisedBy(gameIdentity = command.gameIdentity), 500 millis)
    val gameInitialisedUpdated = reorderPlayerCards(gameInitialised, command.playerIdentity, command.orderedCards)
    gameRepositoryPort.save(gameInitialisedUpdated)
  }

  private def reorderPlayerCards(game: BurracoGameInitiated, playerIdentity: PlayerIdentity, orderedCards: List[Card]) = {
    game match {
      case game: BurracoGameInitiatedTurnStart => game.updatePlayerCardsOrder(playerIdentity, orderedCards)
      case game: BurracoGameInitiatedTurnExecution => game.updatePlayerCardsOrder(playerIdentity, orderedCards)
      case game: BurracoGameInitiatedTurnEnd => throw new Exception("Not yet implemented the possibility to update card order in the turn end phase ") //game.updatePlayerCardsOrder(playerIdentity,orderedCards)
      case _ => throw new Exception
    }
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/OrganisePlayerCardsCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.domainModels.burracoGameExecutions.BurracoGameExecution
import com.abaddon83.burracoGame.domainModels.burracoGameExecutions.BurracoGameExecutionTurnBeginning
import com.abaddon83.burracoGame.domainModels.burracoGameExecutions.BurracoGameExecutionTurnEnd
import com.abaddon83.burracoGame.domainModels.burracoGameExecutions.BurracoGameExecutionTurnExecution
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.decks.Card
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.burracoGame.shared.players.PlayerIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class OrganisePlayerCardsCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        val orderedCards: List<Card>,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity,
                playerIdentity: PlayerIdentity,
                orderedCards: List<Card>): this(
            gameIdentity = gameIdentity,
            playerIdentity = playerIdentity,
            orderedCards = orderedCards,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis())
    )
}

class OrganisePlayerCardsHandler() : CommandHandler<OrganisePlayerCardsCmd>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(command: OrganisePlayerCardsCmd) {
        val gameExecution = runBlocking { repository.findBurracoGameExecutionBy(gameIdentity = command.gameIdentity) }
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameExecution = reorderPlayerCards(gameExecution, command.playerIdentity, command.orderedCards)
        repository.save(updatedGameExecution)
    }

    override suspend fun handleAsync(command: OrganisePlayerCardsCmd) {
        val gameExecution = repository.findBurracoGameExecutionBy(gameIdentity = command.gameIdentity)
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameExecution = reorderPlayerCards(gameExecution, command.playerIdentity, command.orderedCards)
        repository.save(updatedGameExecution)
    }

    private fun reorderPlayerCards(game: BurracoGameExecution, playerIdentity: PlayerIdentity, orderedCards: List<Card>):BurracoGameExecution {
       return when(game){
           is BurracoGameExecutionTurnBeginning -> game.updatePlayerCardsOrder(playerIdentity, orderedCards)
           is BurracoGameExecutionTurnExecution -> game.updatePlayerCardsOrder(playerIdentity, orderedCards)
           is BurracoGameExecutionTurnEnd -> throw Exception("Not yet implemented the possibility to update card order in the turn end phase ") //game.updatePlayerCardsOrder(playerIdentity,orderedCards)
           else -> throw Exception()
        }
    }
}